### PR TITLE
Add integration test for authd remote enrollment option

### PR DESCRIPTION
**Pull request by spothound**
|Related issue|
|---|
|Closes https://github.com/wazuh/wazuh-qa/issues/1326|

## Description

- An added a new test called test_remote_enrollment.py
- Check that remote enrollment works when it is enabled in the configuration.
- Check that remote enrollment fails (connection refused) when trying to connect to 1515 socket if it is disabled on configuration.
- Check that logs `Port 1515 was set as disabled` or `Accepting connections on port 1515. No password required.` appear depending on the configuration.
- Add support for cluster worker and master configuration as well as stand-alone server configuration. 
- Manual testing on local environment.

image

- Add documentation for the new test

## Configuration options
```xml
<auth>
  <remote_enrollment>yes</remote_enrollment>
</auth>
```
## Logs example

```
Installed /usr/local/lib/python3.8/dist-packages/wazuh_testing-4.3.0-py3.8.egg
Processing dependencies for wazuh-testing==4.3.0
Finished processing dependencies for wazuh-testing==4.3.0
root@focal:/tmp/wazuh-qa/1326-auth-options/deps/wazuh_testing# tail -f  /var/ossec/logs/ossec.log 
2021/05/25 13:23:25 wazuh-authd: INFO: Started (pid: 121373).
2021/05/25 13:23:25 wazuh-authd: INFO: Accepting connections on port 1515. No password required.
2021/05/25 13:23:25 wazuh-authd: INFO: Setting network timeout to 1.000000 sec.
2021/05/25 13:23:26 wazuh-authd: INFO: New connection from 127.0.0.1
2021/05/25 13:23:26 wazuh-authd: ERROR: SSL Error (0)
2021/05/25 13:23:26 wazuh-authd: INFO: New connection from 127.0.0.1
2021/05/25 13:23:26 wazuh-authd: INFO: Received request for a new agent (user1) from: 127.0.0.1
2021/05/25 13:23:26 wazuh-authd: INFO: Duplicated name 'user1' (037). Removing old agent.
2021/05/25 13:23:26 wazuh-authd: INFO: Agent key generated for 'user1' (requested by any)
2021/05/25 13:23:26 wazuh-authd: INFO: (1225): SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
2021/05/25 13:23:27 wazuh-authd: INFO: Exiting...
tail: /var/ossec/logs/ossec.log: file truncated
2021/05/25 13:23:27 wazuh-authd: INFO: Started (pid: 121398).
2021/05/25 13:23:27 wazuh-authd: INFO: Port 1515 was set as disabled.
2021/05/25 13:23:29 wazuh-authd: INFO: (1225): SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
2021/05/25 13:23:30 wazuh-authd: INFO: Exiting...
tail: /var/ossec/logs/ossec.log: file truncated
2021/05/25 13:23:30 wazuh-authd: INFO: Started (pid: 121423).
2021/05/25 13:23:30 wazuh-authd: INFO: Port 1515 was set as disabled.
2021/05/25 13:23:31 wazuh-authd: INFO: (1225): SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
2021/05/25 13:23:32 wazuh-authd: INFO: Exiting...
tail: /var/ossec/logs/ossec.log: file truncated
2021/05/25 13:23:32 wazuh-authd: INFO: Started (pid: 121443).
2021/05/25 13:23:32 wazuh-authd: INFO: Accepting connections on port 1515. No password required.
2021/05/25 13:23:32 wazuh-authd: INFO: Setting network timeout to 1.000000 sec.
2021/05/25 13:23:32 wazuh-remoted: INFO: (1409): Authentication file changed. Updating.
2021/05/25 13:23:32 wazuh-remoted: INFO: (1410): Reading authentication keys file.
2021/05/25 13:23:34 wazuh-authd: INFO: New connection from 127.0.0.1
2021/05/25 13:23:34 wazuh-authd: ERROR: SSL Error (0)
2021/05/25 13:23:34 wazuh-authd: INFO: New connection from 127.0.0.1
2021/05/25 13:23:34 wazuh-authd: INFO: Received request for a new agent (user1) from: 127.0.0.1
2021/05/25 13:23:34 wazuh-authd: INFO: Duplicated name 'user1' (038). Removing old agent.
2021/05/25 13:23:34 wazuh-authd: INFO: Agent key generated for 'user1' (requested by any)
2021/05/25 13:23:34 wazuh-authd: INFO: (1225): SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
2021/05/25 13:23:35 wazuh-authd: INFO: Exiting...
tail: /var/ossec/logs/ossec.log: file truncated
2021/05/25 13:23:35 wazuh-authd: INFO: Started (pid: 121469).
2021/05/25 13:23:35 wazuh-authd: INFO: Port 1515 was set as disabled.
2021/05/25 13:23:37 wazuh-authd: INFO: (1225): SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
2021/05/25 13:23:37 wazuh-authd: INFO: Exiting...
tail: /var/ossec/logs/ossec.log: file truncated
2021/05/25 13:23:37 wazuh-authd: INFO: Started (pid: 121489).
2021/05/25 13:23:37 wazuh-authd: INFO: Accepting connections on port 1515. No password required.
2021/05/25 13:23:37 wazuh-authd: INFO: Setting network timeout to 1.000000 sec.
2021/05/25 13:23:38 wazuh-authd: INFO: New connection from 127.0.0.1
2021/05/25 13:23:38 wazuh-authd: ERROR: SSL Error (0)
2021/05/25 13:23:38 wazuh-authd: INFO: New connection from 127.0.0.1
2021/05/25 13:23:38 wazuh-authd: INFO: Received request for a new agent (user1) from: 127.0.0.1
2021/05/25 13:23:38 wazuh-authd: INFO: Duplicated name 'user1' (039). Removing old agent.
2021/05/25 13:23:38 wazuh-authd: INFO: Agent key generated for 'user1' (requested by any)
2021/05/25 13:23:39 wazuh-authd: INFO: (1225): SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
2021/05/25 13:23:40 wazuh-authd: INFO: Exiting...
tail: /var/ossec/logs/ossec.log: file truncated
2021/05/25 13:23:40 wazuh-authd: INFO: Started (pid: 121515).
2021/05/25 13:23:40 wazuh-authd: INFO: Port 1515 was set as disabled.
2021/05/25 13:23:42 wazuh-remoted: INFO: (1409): Authentication file changed. Updating.
2021/05/25 13:23:42 wazuh-remoted: INFO: (1410): Reading authentication keys file.
2021/05/25 14:03:17 wazuh-modulesd:syscollector: INFO: Starting evaluation.
2021/05/25 14:03:17 wazuh-modulesd:syscollector: INFO: Evaluation finished.
```

## Tests

- [x] Proven that tests **pass** when they have to pass.
- [x] Proven that tests **fail** when they have to fail.
- [x] Python codebase satisfies PEP-8 style style guide. `pycodestyle --max-line-length=120 --show-source --show-pep8 file.py`.
- [x] Python codebase is documented following the Google Style for Python docstrings.
- [x] The test is documented in wazuh-qa/docs.
- [x] `provision_documentation.sh` generate the docs without errors.
image
image
image

### CODE REVIEW

File: deps/wazuh_testing/wazuh_testing/tools/sockets.py [hunk 1]
```diff
@@ -50,3 +51,24 @@ def send_active_response_message(active_response_command):
...
+        TimeoutError: The port isn't accepting connection after time specified in `timeout`.
+    """
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    start_time = time.time()
+    while time.time() - start_time < timeout:
+        try:
+            sock.connect((host, port))
+            return
+        except ConnectionRefusedError:
+            time.sleep(1)
```
Comment (BraulioV) [inline deps/wazuh_testing/wazuh_testing/tools/sockets.py:72]:
```suggestion
        try:
            sock.connect((host, port))
            return
        except ConnectionRefusedError:
            time.sleep(1)
        finally:
            sock.close()
```

Don't forget to close the socket if it's no longer necessary.

File: tests/integration/test_authd/test_remote_enrollment.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+
+def test_remote_enrollment(get_configuration, configure_environment, restart_authd):
+    """Check that Authd remote enrollment is enabled/disabled according to the configuration.
+
+    By default, remote enrollment is enabled. When disabled, Authd TLS port (1515 by default) won't be listening
+    to new connection, but requests to local socket will still be attended.
+
+    Raises:
+
```
Comment (BraulioV) [inline tests/integration/test_authd/test_remote_enrollment.py:68]: The docs is missing.

File: tests/integration/test_authd/test_remote_enrollment.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+]
+
+test_data_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'data')
+configurations_path = os.path.join(test_data_path, 'wazuh_conf.yaml')
+configurations = load_wazuh_configurations(configurations_path, __name__, params=parameters, metadata=metadata)
+
+# Variables
+log_monitor_paths = []
+
+receiver_sockets_params = [(("localhost", 1515), 'AF_INET', 'SSL_TLSv1_2')]
```
Comment (BraulioV) [inline tests/integration/test_authd/test_remote_enrollment.py:47]: It would be interesting to change the port to a different one. Just in case.

**Review by BraulioV — CHANGES_REQUESTED** [review]

File: deps/wazuh_testing/wazuh_testing/tools/sockets.py (hunk 1 above)
Comment (spothound) [reply deps/wazuh_testing/wazuh_testing/tools/sockets.py:72]: This causes an error sometimes when the finally is executed and the socket isn't actually connected:

```
OSError: [Errno 9] Bad file descriptor
```

But I have added the close statement at the end of the try block.

File: tests/integration/test_authd/test_remote_enrollment.py (hunk 3 above)
Comment (spothound) [reply tests/integration/test_authd/test_remote_enrollment.py]: https://github.com/wazuh/wazuh-qa/pull/1368/commits/a283be84da01e3d068feb6651acd9c5af6b1606f

File: tests/integration/test_authd/test_remote_enrollment.py (hunk 2 above)
Comment (spothound) [reply tests/integration/test_authd/test_remote_enrollment.py]: https://github.com/wazuh/wazuh-qa/pull/1368/commits/a283be84da01e3d068feb6651acd9c5af6b1606f

**Review by BraulioV — APPROVED** [review]

**Pull request closed without merging**